[Feature] Enhanced MaskEditor to an Image Canvas

brucew4yn3rp opened the pull request:
@trsommer

Revamped the MaskEditor UI and changed it to an Image Canvas. The context menu name has also been renamed accordingly.

Cleaner PR of [2921](https://github.com/Comfy-Org/ComfyUI_frontend/pull/2921)

The interface now supports drawing on the image itself (on the image layer) in addition to covering the area with a mask. Users can select a color or use an eyedropper tool for color matching. This is a powerful tool to guide the image generation process post-masking (e.g., inpainting). Speaking with ComfyUI users for months, this has been highly desired addition.

The tools on the left sidebar have been updated accordingly. The top one, is the existing mask brush, and has a matching mask logo. The second one is the image painting brush.

![image](https://github.com/user-attachments/assets/190d1a44-26fc-4ef4-b9fc-f4de2d17936a)

![image](https://github.com/user-attachments/assets/00de10d5-08a3-4b25-a9a0-c425cf7d457c)

Some relevant context from the old PR

> I need a way to only erase the drawing or the mask part. The easiest solution would be to make the layer clickable to select which layer to work on.

Layer buttons have been implemented. Also, right click will only apply to erasing whichever tool is activated. Additionally, the eraser tool will only erase on the active layer. Selecting the mask tool automatically activates the mask layer, and selecting the Paint tool automatically activates the Paint layer.

Thank you very much for your help in getting this over the line @duckcomfy. I had some issues with the prior PR so decided to close it and initiate a new, cleaner PR.

brucew4yn3rp committed:
Changed the MaskEditor to an Image Canvas

brucew4yn3rp committed:
Updated Image Canvas to properly handle painting and mask on separate layers

brucew4yn3rp committed:
Removed Sortable.js

brucew4yn3rp committed:
Revised Image Canvas Functionality; Fixed Bus; Pending RGB Canvas Save Functionality

brucew4yn3rp committed:
Ran NPX Prettier and minor edits to maskeditor.ts

brucew4yn3rp committed:
Painted pixels are now saved in a combined as well as separate image. Still need to figure out how to load the painted pixels onto RGB Canvas when the MaskEditor is re-opened after painting and saving

brucew4yn3rp committed:
Painted pixels are now editable. Still need to figure out how to pass them as part of the image. Maybe a combinedCanvas?

brucew4yn3rp committed:
Fixed some errors and merged MaskEditor from @duckcomfy

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2026

brucew4yn3rp committed:
Fixed issues that were preventing build

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp commented:
@christian-byrne @webfiltered Tagging you in case you have a chance to take a look.

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

trsommer commented:
I just tried it. There seems to be a problem where the resulting layers are still split into 2 images in the load image preview. I tried it in both chrome and firefox and just painting on the canvas with the paint tool results in 2 images in the load image node
<img width="1360" alt="image" src="https://github.com/user-attachments/assets/ecbcefdb-cc44-4395-bc0c-bf86c98921f5" />
<img width="1360" alt="image" src="https://github.com/user-attachments/assets/d25c280a-ed30-4615-843b-719e86ac0d72" />

after reloading, it seems to be combined:
<img width="1360" alt="image" src="https://github.com/user-attachments/assets/0010e638-955c-4e6c-b9fc-ebd1ce3d41fb" />

Also, if you run the workflow, the correct image seems to be handed over. The core process seems to work, but the previews in the load image node are wrong

UI fixes, not that important:
when the user selects the mask or paint tool, auto select the mask or image layer. It does not make sense for the user paint/mask without a selected layer

trsommer reviewed (changes requested):
I just tried it. There seems to be a problem where the resulting layers are still split into 2 images in the load image preview. I tried it in both chrome and firefox and just painting on the canvas with the paint tool results in 2 images in the load image node
<img width="1360" alt="image" src="https://github.com/user-attachments/assets/ecbcefdb-cc44-4395-bc0c-bf86c98921f5" />
<img width="1360" alt="image" src="https://github.com/user-attachments/assets/d25c280a-ed30-4615-843b-719e86ac0d72" />

after reloading, it seems to be combined:
<img width="1360" alt="image" src="https://github.com/user-attachments/assets/0010e638-955c-4e6c-b9fc-ebd1ce3d41fb" />

If you run the workflow, the correct image seems to be handed over. The core process seems to work, just the previews in the load image node are wrong

UI fixes, not that important:
when the user selects the mask or paint tool, auto select the mask or image layer. It does not make sense for the user paint/mask without a selected layer

brucew4yn3rp commented:
Thanks @trsommer. I think with the current infrastructure across the frontend, for the user to be able to edit the painting after closing out and re-opening the interface, the node has to have the images separate. 

@duckcomfy Changing the rendering of the LoadImage node doesn't seem to work well. Let me know if you have any other ideas.

duckcomfy commented:
Thank you for your quick review @trsommer.
@brucew4yn3rp Please note, if it helps understanding the issue at hand, that when refreshing the page, although the correct preview is displayed, the paint layer is no longer editable.
I'm going to have a look and try to understand how the Load Image node knows what to preview, how the Mask Editor loads the different layers, and why the Load Image node contents are different when refreshing the page. If you have any insight to share on this topic please do.

Commit:
fix: show correct preview in Load Image node after using MaskEditor to paint an image

asagi4 commented:
I gave this a quick try as well. Seems to work well enough and definitely solves a pain point with ComfyUI.

I did find a bug: the brush shape seems to be stuck to circular. If I select the rectangular brush, it'll show a rectangular outline but still make circular marks.

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

Commit:
feat: refine layer-switching UX

Also comes with a refactor of the associated code to be more DRY and maintainable.

In order to explain the new UX let me first explain what I'll call the different tools, as the names we previously used are now ambiguous.

- "Mask Pen": Previously named "Pen" internally, this is an existing tool lets you paint a mask on the image.
- "Paint Pen": New tool which lets you paint colored pixels on the image.
- "Eraser": Exising tool which lets you erase parts of your mask. As of @brucew4yn3rp's PR, you can also erase parts of your painted pixels, depending on which layer is active.
- "Mask Bucket": Previously named "Paint Bucket" internally, this is an existing tool which lets you take a mask that's just an outline, and fill in the inside of your mask outline with a full masked area. Equivalent functionality for the Paint layer is NOT implemented.
- "Mask Color Fill": Previously named "Color Select" internally, this is an existing tool which can turn pixels into a masked area, based on color similarity in contiguous pixels on the base image. Equivalent functionality for the Paint layer is NOT implemented.

New UX:

- The "Mask" layer is now selected when you open the Mask Editor, instead of no layer selected.
- Selecting the "Mask Pen", "Mask Bucket", or "Mask Color Fill" tool automatically selects the Mask layer.
- Selecting the "Paint Pen" tool automatically selects the Paint layer.
- Selecting the "Mask" layer while the "Paint Pen" tool is selected automatically selects the "Mask Pen" tool.
- Selecting the "Paint" layer while the "Mask Pen", "Mask Bucket", or "Mask Color Fill" tool is selected automatically selects the "Paint Pen" tool.
- Changing the selected layer while the "Eraser" tool is selected has no additional effect, because Eraser supports both layers.
- Switching to the "Eraser" tool has no additional effect, because Eraser supports both layers.

Commit:
fix: make rectangular brush shape work again

Commit:
feat: `[` and `]` MaskEditor shortcuts to decrease/increase brush size

brucew4yn3rp committed:
Merge pull request #2 from duckcomfy/feat/image_canvas_new_pr

fix: show correct preview after painting an image in MaskEditor
improve layer-selecting UX
fix rectangular brushes
add keyboard shortcuts to change brush size

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp commented:
Huge thanks to @duckcomfy for taking the charge with smoothing the rough edges that were remaining with this PR:

- Ensures the node preview of a loaded image is that of the combined canvas (base image + mask + paint), with no other previewable image. [3e259c9](https://github.com/brucew4yn3rp/ComfyUI_frontend_ImageEditor/pull/2/commits/3e259c9ea60014bf5b3113063f63100674340857)
- Improves layer-switching UX (details below) [fbff95b](https://github.com/brucew4yn3rp/ComfyUI_frontend_ImageEditor/pull/2/commits/fbff95b8eac8d9cbeff8da5be6e570719dfbded4)
- Fixes the regression where square brush shapes became always round (NOTE: when hardness<1, brush shape becomes round, but this is true in main branch as well) [118522f](https://github.com/brucew4yn3rp/ComfyUI_frontend_ImageEditor/pull/2/commits/118522f1c9a2056ed3d758296b37dcb71d469286)
- Adds [ and ] as keyboard shortcuts to decrease/increase brush size [c744eeb](https://github.com/brucew4yn3rp/ComfyUI_frontend_ImageEditor/pull/2/commits/c744eeb774083bedbf69778fd6ac60350521376d)

## Preview image bug and fix explanation

The previewed images are pulled from the `LGraphNode` object's `imgs` (`HTMLImageElement[]`) property, and the way the code works, MaskEditor is initialized expecting the different layers to also be in that `imgs` array, and refactoring that would be difficult and high-risk. As a simple fix, I added a `data-nopreview` data attribute to the HTMLImageElement objects, to be able to supply MaskEditor with the discrete layers while only previewing the combined canvas.

~~Before the main repo PR can be merged, we will also need to automatically select an image layer when opening the MaskEditor.~~ Fixed in 5187121c6160961a76cefa3252c2eb4ad4ec8880

Note: after refreshing the page, the paint layer is no longer editable. I'd suggest not considering this bug as a blocker for the main repo PR, as it could be a deep rabbit hole for a very minor bug.

## Layer UX improvements

This PR improves the layer UX in [fbff95b](https://github.com/brucew4yn3rp/ComfyUI_frontend_ImageEditor/pull/2/commits/fbff95b8eac8d9cbeff8da5be6e570719dfbded4). Here is an explanation copied from the commit of the UX improvements. We should copy this explanation to the parent PR after this child PR, for the benefit of reviewers.

In order to explain the new UX let me first explain what I'll call the different tools, as the names we previously used are now ambiguous.

- "Mask Pen": Previously named "Pen" internally, this is an existing tool lets you paint a mask on the image.
- "Paint Pen": New tool which lets you paint colored pixels on the image.
- "Eraser": Exising tool which lets you erase parts of your mask. As of @brucew4yn3rp's PR, you can also erase parts of your painted pixels, depending on which layer is active.
- "Mask Bucket": Previously named "Paint Bucket" internally, this is an existing tool which lets you take a mask that's just an outline, and fill in the inside of your mask outline with a full masked area. Equivalent functionality for the Paint layer is NOT implemented. **I'd suggest this not being a blocker for the main PR.**
- "Mask Color Fill": Previously named "Color Select" internally, this is an existing tool which can turn pixels into a masked area, based on color similarity in contiguous pixels on the base image. Equivalent functionality for the Paint layer is NOT implemented. **I'd suggest this not being a blocker for the main PR.**

New UX:

- The "Mask" layer is now selected when you open the Mask Editor, instead of no layer selected.
- Selecting the "Mask Pen", "Mask Bucket", or "Mask Color Fill" tool automatically selects the Mask layer.
- Selecting the "Paint Pen" tool automatically selects the Paint layer.
- Selecting the "Mask" layer while the "Paint Pen" tool is selected automatically selects the "Mask Pen" tool.
- Selecting the "Paint" layer while the "Mask Pen", "Mask Bucket", or "Mask Color Fill" tool is selected automatically selects the "Paint Pen" tool.
- Changing the selected layer while the "Eraser" tool is selected has no additional effect, because Eraser supports both layers.
- Switching to the "Eraser" tool has no additional effect, because Eraser supports both layers.

![1751908337](https://github.com/user-attachments/assets/98a92e03-b2de-44df-893c-3042540b3c00)

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

christian-byrne commented:
I can review if needed but will defer to @trsommer!

trsommer commented:
Thank you for addressing the previous issues! I’ve tested the updated version, and it’s very close to perfect - great work! I did notice one small issue: When a mask overlaps a painted area and the user reopens the mask editor, the mask is subtracted from the painted area.

<img width="1360" alt="image" src="https://github.com/user-attachments/assets/62b94c70-1820-4e71-a69b-fbeb45dc02a2" />

reopen the mask editor and save

<img width="1360" alt="image" src="https://github.com/user-attachments/assets/1da89932-3f7c-4d00-8fa0-dbf1a828e140" />

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

Commit:
refactor: run prettier

Commit:
fix: no longer erase intersection between paint and mask in paint layer when re-opening mask editor

nit: clean up some naming and stray console logs

brucew4yn3rp committed:
Merge pull request #3 from duckcomfy/fix/paint_mask_overlap

fix: no longer erase intersection between paint and mask in paint layer when re-opening mask editor

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp commented:
Thanks again for the quick review @trsommer. 

@duckcomfy fixed the issue in the latest PR I’ve just merged in by correcting the ordering in which each layer was being pushed to the imgs array. Also ran prettier to align with formatting standard.

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

trsommer reviewed (dismissed):
thank you for fixing the issues I mentioned

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

brucew4yn3rp commented:
Seems like we are in the final stage of introducing a more comprehensive image editing canvas to ComfyUI! 

@christian-byrne As @trsommer has completed his review, would you mind taking a look as well? I think we need a code owner review from comfy_frontend_devs before it can be merged in.

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

webfiltered reviewed (changes requested):
Keen to take a look at this properly!  Looks like a great idea.

Have paused my review partway through though - found the `tsconfig` changes.  Once the below have been addressed, please ping~

Commit:
refactor: re-enable noUnusedLocals, noUnusedParameters, and fix resulting errors

Commit:
fix: undo unnecessary vue-tsc update

Commit:
refactor: tighten up code

brucew4yn3rp committed:
Merge pull request #4 from duckcomfy/refactor/address_review

refactor/address review

brucew4yn3rp committed:
Improved the Layer UX. Re-organized the interface and made it so that 'Activate Layer' buttons are only visible when the Eraser tool is selected, for which they are used to discern which layer is being erased. Otherwise, the buttons are hidden and a blue border reflects which layer is currently active.

brucew4yn3rp committed:
Ran Prettier

brucew4yn3rp committed:
Reverted sidePanel sizing

Commit:
fix: allow maskeditor right-pane to have a scrollbar on small viewports & improve the padding consistency when that happens

brucew4yn3rp committed:
Merge pull request #5 from duckcomfy/Improved-Layer-UX_refinements

fix: allow maskeditor right-pane to have a scrollbar on small viewports & improve the padding consistency when that happens

brucew4yn3rp committed:
Fixed curly brace style nit

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

webfiltered reviewed (changes requested):
Looks pretty good!  Have some follow-ups but nothing major.  The biggest is a comment on `app.ts` about `node.imgs`.

Have also added suggestions which you can just click commit on if that works for you.  If any tests fail because of them, feel free to ping and I can resolve.

Commit:
refactor: address all review items except node.imgs

Commit:
refactor: await http request to avoid race conditions

Commit:
chore: revert to main branch package-lock.json

Commit:
refactor: separate maskeditor layer data from node imgs (previewed images)

Commit:
Revert "refactor: separate maskeditor layer data from node imgs (previewed images)"

This reverts commit c6eb0ba2042302bc4b6d035ecfc08b1ee3b5cee2.

Commit:
refactor/fix: no longer store mask layers in node.imgs, instead derive their url from the main image url, which also ensures the layers can be opened even if the page is refreshed

duckcomfy commented on src/scripts/app.ts:
```diff
@@ -448,6 +479,10 @@ export class ComfyApp {
         }
       }
 
+      node.imgs?.forEach((imgEl) => {
+        if (imgEl.src === combinedImgSrc) return
+        imgEl.setAttribute('data-nopreview', 'true')
+      })
```
At first, I thought the `node` object was the only sensible place to get layer data from, and then the `node.imgs` the only semi-sensible existing property to store that in. After your comment, I first tried an approach that extends LGraphNode (after I noticed the frontend codebase was already doing that) to have a new `maskEditorLayers` prop, but then I realised that the only approach that'll work completely is to derive the layer URLs from the main image URL. The reason why is that this approach survives page refreshes, unlike storing layer data in the node object.

You can see how I implemented this new approach in: [a2ebf51](https://github.com/brucew4yn3rp/ComfyUI_frontend_ImageEditor/pull/6/commits/a2ebf5140273a28a752f42cec2352a7c963ce699) 

I addressed all the other review items in the [stacked PR](https://github.com/brucew4yn3rp/ComfyUI_frontend_ImageEditor/pull/6) which I think @brucew4yn3rp told me on Discord he will merge when he next is available.

asagi4 commented on src/constants/coreKeybindings.ts:
```diff
@@ -181,5 +181,7 @@ export const CORE_KEYBINDINGS: Keybinding[] = [
     commandId: 'Comfy.Graph.ConvertToSubgraph'
-  }
+  },
+  { combo: { key: ']' }, commandId: 'Comfy.MaskEditor.BrushSize.Increase' },
+  { combo: { key: '[' }, commandId: 'Comfy.MaskEditor.BrushSize.Decrease' }
```
Those hotkeys are also not the best for people with non-US keyboard layouts. The brush sizes could use a number input (precise control of size is almost impossible currently) but that's not for this PR

brucew4yn3rp committed:
Merge pull request #6 from duckcomfy/feat/image_editor_review_2

refactor: address review items on main Image Editor PR

Commit:
fix: add back change to ensure node preview is only the combined image

brucew4yn3rp committed:
Merge pull request #7 from duckcomfy/feat/image_editor_review_2

fix: add back change to ensure node preview is only the combined image

brucew4yn3rp commented:
Hi @webfiltered, thanks for the attentiveness on this - I know there's lots going on. 

All of your comments have been addressed in the latest commits. We discussed offline a way to circumvent the need to store images within the node. Therefore, we no longer need the `data-nopreview` workaround either. 

Originally, I had implemented the storing of images in `node.imgs` as it was the only way I'd found to be able to persist which correct files to reference for each layer upon closing and re-opening of the canvas. 

Now, the timestamp has been unified across all the related files, allowing us to derive the names of each layer's saved file from the final filename that gets saved to the node/workflow. As a byproduct, this allows editability to persist through refreshes because deriving the filenames is static (as opposed to `node.imgs` which did not persist through app refresh),

This seems to work well, and without the additional clog in retaining images within the node, which you mentioned may be a strain for future maintainability. To that end, @duckcomfy has also added a `maskeditor.test.ts` file that will assist in ensuring the new functionality does not break in the future.

brucew4yn3rp committed:
Merge branch 'main' into Jun-5-2025

PabloWiedemann reviewed (approved):
LGTM

kevinpaik1 commented:
@brucew4yn3rp once your PR gets merged, would you please write into me at bounty@comfy.org? i'd love to compensate you as a thank you for your significant effort here!

kevinpaik1 commented:
Hi @trsommer — we’d like to propose adding @brucew4yn3rp as a CODEOWNER for the MaskEditor component, based on:

1. The recent refactoring work in this PR
2. He’s expressed interest in helping review future contributions in this area

@christian-byrne mentioned you're currently listed as the CODEOWNER here, so I wanted to check with you. Would you be open to adding @brucew4yn3rp alongside you? Thanks in advance for considering!

avc commented:
I found a bug I think may be related to this implementation https://github.com/Comfy-Org/ComfyUI_frontend/issues/5251. The mask editor is now anti-aliasing even at max hardness, making fills not fill completely and cleanly.

silveroxides commented:
@trsommer
Thank you. This was most likely an oversight and the lack of manual testing led to it not being detected.
I don't know how much @brucew4yn3rp actually uses the mask editor himself when using ComfyUI but I would expect that if someone is set as codeowner for a feature, then they ought to be dogfooding said feature.

brucew4yn3rp commented:
Thank you @trsommer for the quick fix. That was indeed an oversight during the UI redesign. The PR went through multiple iterations covering both new functionalities and the UI (including the sidebar), so given the scope it’s unfortunate but not surprising that this slipped through during the rebuild. Again, given the scope and difficulty of implementing the new functions, I was also admittedly focused on making sure the new features worked well and missed the regression which probably happened early on.

>just that this was so sloppily implemented and completely lacked quality control and even the mere basic manual testing.

@silveroxides, instead of assuming “sloppy implementation” or lack of testing, it would be more constructive to check the PR history, which clearly shows extensive review and collaboration over several months. In a refactor of this size, oversights are inevitable, which is why collaboration exists. If you run into issues in the future, the most helpful approach is to either open an issue or submit a PR. And if you remain convinced this reflects “sloppy” work, you’re always free to continue using the old mask editor.
